### PR TITLE
Implement dispatch logic to inventory each project on a seperate thread.

### PR DESCRIPTION
**Pull request by ahoying**
* Use concurrent.futures to setup a threadpool with 10 workers.
* Dispatch project resources to the threadpool.
* Implement write locking on the storage.

Baseline run time for our test organization:
5m31s

Parallel run time for our test organization using this branch:
1m12s

Data in inventory identical.

Thanks for opening a Pull Request!

Here's a handy checklist to ensure your PR goes smoothly.

- [x] I signed Google's [Contributor License Agreement](https://opensource.google.com/docs/cla/)
- [x] My code conforms to Google's [python style](https://google.github.io/styleguide/pyguide.html).
- [x] My PR at a minimum doesn't decrease unit-test coverage (if applicable).
- [x] My PR has been functionally tested.
- [x] All of the [unit-tests](http://forsetisecurity.org/docs/development/#executing-tests) still pass.
- [x] Running `pylint --rcfile=pylintrc` passes.

These guidelines and more can be found in our [contributing guidelines](https://github.com/GoogleCloudPlatform/forseti-security/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: google/cloud/forseti/services/inventory/crawler.py [hunk 1]
```diff
@@ -12,15 +12,18 @@
...
-from google.cloud.forseti.services.inventory.base import gcp
+import threading
+from concurrent.futures import ThreadPoolExecutor
+
 from google.cloud.forseti.services.inventory.base import crawler
+from google.cloud.forseti.services.inventory.base import gcp
+from google.cloud.forseti.services.inventory.base import resources
 
 
 class CrawlerConfig(crawler.CrawlerConfig):
```
[the remark this answers is not in the document]
Comment (ahoying) [reply google/cloud/forseti/services/inventory/crawler.py]: Done.

File: google/cloud/forseti/services/inventory/crawler.py [hunk 2]
```diff
@@ -116,15 +135,110 @@ def update(self, resource):
...
+        """Process items in the queue until the shutdown event is set."""
+        while not self._shutdown_event.is_set():
+            try:
+                (callback, resource) = self._dispatch_queue.get(timeout=1)
+            except Empty:
+                continue
+
+            try:
+                callback()
+            except Exception as e:
```
Comment (felixbb) [inline google/cloud/forseti/services/inventory/crawler.py]: Is it possible to have this handler in the callback?

The reason I'm asking is that I believe what the handler does is a implementation detail of the resource so it should be part of the callback closure. This way, the crawler just schedules callables and does really not have to care what these callables do.

File: google/cloud/forseti/services/inventory/base/resources.py [hunk 3]
```diff
@@ -124,19 +121,19 @@ def accept(self, visitor, stack=None):
...
                 for resource in yielder.iter():
                     res = resource
+                    new_stack = stack + [self]
 
-                    def call_accept():
-                        res.accept(visitor, stack + [self])
-
-                    if res.is_leaf():
-                        call_accept()
+                    if res.type() != 'project':
```
Comment (felixbb) [inline google/cloud/forseti/services/inventory/base/resources.py]: Nit: If we could make this a res.should_dispatch() with a base implementation of "return False" and a Project class implementation of "return True", that would be better documentation of what's going on.

Comment (ahoying) [reply google/cloud/forseti/services/inventory/base/resources.py]: Fixed.

File: google/cloud/forseti/services/inventory/crawler.py (hunk 2 above)
Comment (ahoying) [reply google/cloud/forseti/services/inventory/crawler.py]: Fixed.

**Comment by ahoying** [comment]
PTAL.

**Pull request merged**